onnxruntime/core/optimizer/matmul_bn_fusion.cc

```diff
@@ -0,0 +1,280 @@
+#include "core/optimizer/matmul_bn_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+
+namespace onnxruntime
+{
+void AddNodesToRemove(Node::NodeConstIterator current_iterator,
+                      const NodeIndex& dest_node_index,
+                      std::vector<NodeIndex>& nodes_to_remove) {
+  while (current_iterator->Index() != dest_node_index) {
+    nodes_to_remove.push_back(current_iterator->Index());
+    current_iterator = current_iterator->OutputNodesBegin();
+  }
+}
+
+NodeIndex GetOtherParentOfNode(const Node& node, NodeIndex first_parent_index) {
+  NodeIndex other_parent_index = std::numeric_limits<size_t>::max();
+  if (node.GetInputEdgesCount() != 2) {
+    return other_parent_index;
+  }
+
+  auto parent_node_iterator = node.InputNodesBegin();
+  if (parent_node_iterator->Index() != first_parent_index) {
+    other_parent_index = parent_node_iterator->Index();
+  }
+  ++parent_node_iterator;
+  if (parent_node_iterator->Index() != first_parent_index) {
+    other_parent_index = parent_node_iterator->Index();
+  }
+  return other_parent_index;
+}
+
+bool MatchPath(const Node& parent_node,
+               const gsl::span<std::pair<std::string, InlinedVector<ONNX_NAMESPACE::OperatorSetVersion>>>& path,
+               const Node& child_node) {
+  if (path.size() == 0) {
+    return true;
+  }
+
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(child_node, path[0].first, path[0].second) ||
+      child_node.GetExecutionProviderType() != parent_node.GetExecutionProviderType()) {
+    return false;
+  }
+
+  /*
+  * last node in the path can have more than one output
+  * because all those outputs will be preserved by the addition of new Gemm node
+  */
+  if (path.size() > 1 && child_node.GetOutputEdgesCount() != 1) {
+    return false;
+  }
+
+  return MatchPath(child_node, path.subspan(1), *child_node.OutputNodesBegin());
+}
+
+/*
+*   Given a MatMul node, it will verify the following pattern.
+*                      MatMul
+*                        |
+*                       / \
+*                      /   \
+*                     /     \
+*               Reshape     Shape
+*                  |          |
+*             Transpose      Cast
+*                  |          |
+*        BatchNormalization  Cast
+*                  |          |
+*              Transpose      |
+*                  |         /
+*                   \       /
+*                    \     /
+*                     \   /
+*                       | 
+*                    Reshape
```

Comment:
I would expect that Shape->Cast->Cast to be eliminated by earlier iterations of constant folding.  This would mean that they don't need to be handled here, and in fact that we should be robust to those nodes being missing.

More generally, the smaller we make the pattern detected in this individual pass, the more the optimization will apply to other models *and* be robust to changes and improvements in upstream optimization steps.

This should also be able to rely on Transpose->Transpose elimination and Reshape elimination (at least for DML EP). Therefore, could this pass simply optimize the sequence below?

(MatMul-> [Reshape] -> [Transpose] -> BatchNorm) ---> GEMM -> [Reshape] -> [Transpose]

Reply:
Makes sense. The update pattern matching works too. And `ConstantFolding` L1 optimizer removes the Shape -> Cast -> Cast path.